### PR TITLE
Resolves #246: Cached file shall retain the original resource name

### LINKED ISSUE #246: 1.7.0 is broken
- if .cache/download-maven-plugin exists and artifact is deleted from there, plugin fails to work with 1.7.0.  Whereas 1.6.8 will redownload when missing.

```
[ERROR] Failed to execute goal com.googlecode.maven-download-plugin:download-maven-plugin:1.7.0:wget (default) on project apache-ant: General error: : java.nio.file.NoSuchFileException:...\.m2\repository\.cache\download-maven-plugin\apache-ant-1.10.13-bin.tar.gz_c5f78007b556836b518ca981366e5928
```

- if cache folder is deleted, plugin continually states its downloading and never creates cache.  Dropping back to 1.6.8 will re-create.

- Have another example not shown here that doesn't even attempt to download, gets 2 0 byte attempts and creates a 0 byte entry.  Using wget, not sure if because tar.gz requested or if from github or not.  Dropping to 1.6.8 just works.

Configuration, none, out of the box setup.  Simply using 'wget' option and location to download in this case.  On windows.

You can use https://github.com/hazendaz/ant to reproduce.  Its on 1.7.0 now and broken.  Drop to 1.6.8 and it works.

Maven version in use 3.9.1
Java version in use 17.0.6

Reviewed release notes on this project, nothing mentioned about having to make any changes.

### PR DESCRIPTION
**Pull request by ajarmoniuk**
Cached file name will retain the original resource name (like in the old implementation)

@longtimeago please check although it's a small change.

In the previous impl, I assumed (wrongly) that the uri provided by HttpClient would contain the host name, but it doesn't. It's a relative uri rooted at resource root.

### CODE REVIEW

File: src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java [hunk 1]
```diff
@@ -32,9 +30,11 @@ public FileIndexResourceFactory(final Path cacheDir) {
...
 
     protected Path generateUniqueCachePath(final String uri) {
-        Matcher domainMatcher = DOMAIN_REGEX.matcher(uri);
-        String domainPrefix = domainMatcher.find() ? domainMatcher.group(1) + '_' + DigestUtils.md5Hex(uri) : "";
-        return Paths.get(domainPrefix + DigestUtils.md5Hex(uri));
+        int slashIndex = uri.lastIndexOf('/');
+        String fileName = slashIndex < uri.length() - 1
+                ? uri.substring(slashIndex + 1) + "_"
+                : "";
+           return Paths.get(fileName + DigestUtils.md5Hex(uri));
```
[the remark this answers is not in the document]
Comment (ajarmoniuk) [reply src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java:37]: I know, bot, but this is ain't used for any cryptography.

**Comment by ajarmoniuk** [comment]
Ran into some problems:
- turns out, the URI provided by Apache HttpClient Cache is not absolute and does not contain the host name at all, had to find another uniqueness factor for the cache file
- WGetMojo would barf because it wouldn't be able to retrieve the file name if the requested resource is the root resource (e.g. https://www.google.com). So in this case we'll just use the host name as file name

I'll try to resolve these when I have some time later today.

File: src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java [hunk 2]
```diff
@@ -32,20 +28,26 @@ public FileIndexResourceFactory(final Path cacheDir) {
...
-        String domainPrefix = domainMatcher.find() ? domainMatcher.group(1) + '_' + DigestUtils.md5Hex(uri) : "";
-        return Paths.get(domainPrefix + DigestUtils.md5Hex(uri));
+        String resourcePart = !uri.isEmpty()
+                ? uri.substring(uri.lastIndexOf('/') + 1)
+                : uri;
+        resourcePart = resourcePart.isEmpty()
+                ? resourcePart
+                : resourcePart + "_";
+        // append a unique string based on timestamp
+        return Paths.get(resourcePart +
```
[the remark this answers is not in the document]
Comment (ajarmoniuk) [reply src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java:38]: @sonatype-lift ignore

File: src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java [hunk 3]
```diff
@@ -32,20 +28,26 @@ public FileIndexResourceFactory(final Path cacheDir) {
...
+                DigestUtils.md5Hex(String.valueOf(System.currentTimeMillis())));
```
[the remark this answers is not in the document]
Comment (ajarmoniuk) [reply src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java:39]: @sonatype-lift ignore

**Comment by ajarmoniuk** [comment]
Two sonartype findings which were not applicable here: we're not using MD5 for encryption, and we're using Paths.get to construct a path, but not to open a file.

File: src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java [hunk 4]
```diff
@@ -32,20 +28,26 @@ public FileIndexResourceFactory(final Path cacheDir) {
     }
 
     protected Path generateUniqueCachePath(final String uri) {
-        Matcher domainMatcher = DOMAIN_REGEX.matcher(uri);
-        String domainPrefix = domainMatcher.find() ? domainMatcher.group(1) + '_' + DigestUtils.md5Hex(uri) : "";
-        return Paths.get(domainPrefix + DigestUtils.md5Hex(uri));
+        String resourcePart = !uri.isEmpty()
```
[the remark this answers is not in the document]
Comment (longtimeago) [reply src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java:31]: @ajarmoniuk javadoc for the method is also appreciated especially mentioning the concern above

[the remark this answers is not in the document]
Comment (ajarmoniuk) [reply src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java:31]: > btw, with this PR we are effectively invalidating all cache entries enforcing downloading then again on request.

Why?

[the remark this answers is not in the document]
Comment (ajarmoniuk) [reply src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java:31]: > Afak, initially we were needed predictable resource path to be able to construct it later and fetch resource from the cache. 

Not really. The cache index contains that information.

I think the first version of the plugin (up to 1.3.0) relied on the file name containing the hashed url, but this information is now present in the index.

File: src/test/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactoryTest.java [hunk 5]
```diff
@@ -1,28 +1,55 @@
...
 /**
  * Unit tests for {@linkplain FileIndexResourceFactory}
  */
 public class FileIndexResourceFactoryTest {
 
+    private FileIndexResourceFactory factory;
+
+    @Before
+    public void setUp() {
+        this.factory = new FileIndexResourceFactory(Paths.get("/tmp"));
```
[the remark this answers is not in the document]
Comment (ajarmoniuk) [reply src/test/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactoryTest.java:20]: Test cases don't share anything between each other. The factory instance is recreated for every single test.

**Comment by ajarmoniuk** [comment]
> So, why complicate things and not using simple uri + UUID instead?

Frankly, I don't see how UUID is simpler 😉

**Comment by ajarmoniuk** [comment]
Sadly, I don't agree with the review comments. Please provide more justification.

File: src/test/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactoryTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (longtimeago) [reply src/test/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactoryTest.java:20]: @ajarmoniuk I'm not saying tests share instance. But they become logically coupled. Means changing the way factory is instantiated will change that for all tests in file.
Also, I'm a big fan of immutability, and here `this.factory` is non-final instance variable :)

File: src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (longtimeago) [reply src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java:31]: @ajarmoniuk doh! Please, ignore this thread

[the remark this answers is not in the document]
Comment (longtimeago) [reply src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java:31]: > Not really. The cache index contains that information.

@ajarmoniuk doh! Please, ignore suggestion

**Comment by longtimeago** [comment]
> Frankly, I don't see how UUID is simpler

@ajarmoniuk well, ok, if we can't get rid of this https://github.com/maven-download-plugin/maven-download-plugin/pull/251/files#diff-871a20e5bff274a5d2ef96fffd9e4e5afa872da4ae93fc345a3e2157c838d0dfR31-R38 than it's probably does't matter

File: src/main/java/com/googlecode/download/maven/plugin/internal/CommonUtils.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package com.googlecode.download.maven.plugin.internal;
+
+import java.net.URI;
+
+/**
+ * Common utilities used by the {@link WGetMojo}
+ */
+public class CommonUtils {
```
Comment (longtimeago) [inline src/main/java/com/googlecode/download/maven/plugin/internal/CommonUtils.java:8]:
```suggestion
public final class FileNameUtils {
```

**Comment by ajarmoniuk** [comment]
> > Frankly, I don't see how UUID is simpler
> 
> @ajarmoniuk well, ok, if we can't get rid of this https://github.com/maven-download-plugin/maven-download-plugin/pull/251/files#diff-871a20e5bff274a5d2ef96fffd9e4e5afa872da4ae93fc345a3e2157c838d0dfR31-R38 than it's probably does't matter

We only need uniqueness should a resource with the same name be requested using different URIs, then they do need separate file names. This could have been achieved if we had access to the whole (not relative) URI, but can also be achieved using a timestamp. UUID is overkill IMO, merely consulting the system time should do.

**Comment by longtimeago** [comment]
> UUID is overkill IMO, merely consulting the system time should do.

@ajarmoniuk well ... PC time is not monotonic function (thanks to NTP) :)

**Comment by ajarmoniuk** [comment]
> > UUID is overkill IMO, merely consulting the system time should do.
> 
> @ajarmoniuk well ... PC time is not monotonic function (thanks to NTP) :)

I mean, I don't really care. If you insist, we can do e.g.:

1) resourcePart + DigestUtils.md5Hex(UUID.randomUUID().toString())
2) resourcePart + UUID.randomUUID().toString()

or the UUID can be based on the local resource name.

**Comment by longtimeago** [comment]
> 1. resourcePart + DigestUtils.md5Hex(UUID.randomUUID().toString())

@ajarmoniuk let's use this

File: src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java [hunk 7]
```diff
@@ -32,20 +29,26 @@ public FileIndexResourceFactory(final Path cacheDir) {
...
+                ? uri.substring(uri.lastIndexOf('/') + 1)
+                : uri;
+        resourcePart = resourcePart.isEmpty()
+                ? resourcePart
+                : resourcePart + "_";
+        // append a unique string based on timestamp
+        return Paths.get(resourcePart +
+                DigestUtils.md5Hex(UUID.randomUUID().toString()));
```
[the remark this answers is not in the document]
Comment (ajarmoniuk) [reply src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileIndexResourceFactory.java:40]: @sonatype-lift ignore

**Review by longtimeago — APPROVED** [review]
@ajarmoniuk thank you for your patience! :+1:

**Pull request merged**